src/experimental/irregular_buffer.zig

```diff
@@ -0,0 +1,201 @@
+
+const std = @import("std");
+
+const Allocator = std.mem.Allocator;
+
+pub fn IrregularBuffer(comptime T: type) type {
+
+    return struct {
+
+        const Self = @This();
+        const DataSlice = []T;
+        const ConstData = [] const T;
+        const IdxsSlice = []IndexPair;
+        const ConstIdxs = [] const IndexPair;
+
+        const IndexPair = struct {
+            lhs: usize,
+            rhs: usize,  
+        };
+
+        const ForwardIterator = struct {
+            ptr: *const Self,
+            idx: usize = 0,
+            pub fn next(self: *@This()) ?ConstData {
+                const last = self.idx;
+                return if (last < self.ptr.idxs.len) blk: { 
+                    self.idx += 1;
+                    break :blk self.ptr.get(last);
+                } else null;
+            }
+        };
+
+        data: DataSlice,
+        idxs: IdxsSlice,
+        data_capacity: usize,
+        idxs_capacity: usize,
+        allocator: Allocator,
+
+        pub fn init(allocator: Allocator) Self {
+            return Self{
+                .data = &[_]T{},
+                .idxs = &[_]IndexPair{},
+                .data_capacity = 0,
+                .idxs_capacity = 0,
+                .allocator = allocator,
+            };
+        }
+
+        pub fn itemCount(self: *const Self) usize {
+            return self.idxs.len;
+        }
```

Comment:
What do you think about this as a convention?

* `len` -- Collection is linear (array/list/etc) and size is known O(1)
* `size` -- Collection size is statically known O(1)
* `count()` -- Collection size will be dynamically computed at O(n)

Whether `len` or `size` are struct members or functions depends on the details

Reply:
I like that idea! I honestly ran out of synonyms at that point.

So the reason I did itemCount in this case is because I did the same thing in the RegularBuffer (but that one has a good reason to exist... it's calculated on the fly so a division operation has to occur). 

Okay, so in that case... the RegularBuffer should have a count() function, and this one should have a len() function?

Reply:
I think so, and probably we should start a CONVENTIONS.md or something?

Reply:
So here's my proposal:

len - refers to retrieving a data member that continually keeps track of the number of elements

size - refers to retrieving a non-mutable (or statically known) number of elements

count - refers to anything that must calculate the number of elements regardless of complexity (it's literally counting)

Reply:
Oops I never responded... I think that's great